### ISSUE TITLE
Terraform 0.13 cannot init with existing providers in state and no internet

### ISSUE THREAD

**Issue by cmbrad** [hashicorp/terraform#25702]
### Terraform Version

```
Terraform v0.13.0-rc1
...
```

### Terraform Configuration Files

```terraform
terraform {
  backend "s3" {
    kms_key_id = "alias/my-key"
    region = "ap-southeast-2"
    encrypt = true
  }
}
```

### CLI Output

```
Initializing provider plugins...
- Finding latest version of -/aws...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider -/aws: could not connect to registry.terraform.io: Failed to request discovery document: Get "https://registry.terraform.io/.well-known/terraform.json": EOF
```

### Expected Behavior

Terraform inits completes successfully as per 0.12.X

### Actual Behavior

Terraform fails to init

### Steps to Reproduce

1. Be on an instance with AWS API access, but no access to registry.terraform.io / releases.hashicorp.com (no internet)
2. Have an existing state file from 0.12.X in a remote backend which references a provider, for example aws
3. `terraform init`

### Additional Context

The instance has the providers cached to disk in ~/.terraform.d/plugins/registry.terraform.io/hashicorp/aws/2.7.0/linux_amd64/ and ~/.terraform.d/plugins/linux_amd64/

Init completes successfully finding the AWS provider if a new state file is used.

Enabling a http proxy with internet access results in being redirected to download the AWS provider from releases.hashicorp.sun (which then fails due to a DNS lookup because of the server isolation, but anyway...)

**Comment by alisdair** [comment]
Thanks for reporting this issue!

You should be able to fix this issue by updating your statefile to refer to the new provider source. For example:

```shellsession
$ terraform state replace-provider -- -/aws hashicorp/aws
```

Could you try this and report back if it works around the problem?

**Comment by cmbrad** [comment]
Thanks Alisdair! That does work around the problem 👍 It'd be nice if init could magically map -/aws -> hashicorp/aws so we don't need to modify the state to test 0.13 compatibility, although it probably wouldn't be too difficult for us to:
1. Switch to local backend
2. Run state replace
3. Run a test plan

That or just test from an internet-enabled machine. Future problem!

Cheers,
Chris

**Comment by cmbrad** [comment]
In a different workaround I've updated our container build scripts to also write the provider to ~/.terraform.d/plugins/registry.terraform.io/-/aws/2.7.0/linux_amd64/ and now it also works. Although with that it does seem to be treating hashicorp/aws and -/aws as two different providers.

**Comment by GoGoris** [comment]
> Thanks for reporting this issue!
> 
> You should be able to fix this issue by updating your statefile to refer to the new provider source. For example:
> 
> ```
> $ terraform state replace-provider -- -/aws hashicorp/aws
> ```
> 
> Could you try this and report back if it works around the problem?

@alisdair You helped me a lot! Can this be added to the provider documentation of terraform 0.13? https://www.terraform.io/docs/configuration/providers.html

**Comment by danieldreier** [comment]
@GoGoris it will be documented in the [state docs](https://github.com/hashicorp/terraform/blob/9266e944fae5ec78d9c3ccb0bb8f5a1084d23e09/website/docs/commands/state/replace-provider.html.md) when those are published on Monday during the 0.13.0 GA release.

**Comment by danieldreier** [comment]
I'm going to close this issue because the `terraform state replace-provider` command is documented and I think that sufficiently resolves the original issue as stated, but if people want to explicitly file an enhancement request to improve how terraform works without internet I think that's a valid (but larger) request that will take a lot more thinking through. Thanks for reporting this and I'm glad the workaround is working for you!
